drake/solvers/MobyLCP.cpp

```diff
@@ -0,0 +1,1268 @@
+// Adapted with permission from code by Evan Drumwright
+// (https://github.com/edrumwri).
+
+#include "MobyLCP.h"
+
+#include <Eigen/LU>
+#include <Eigen/SparseCore>
+#include <Eigen/SparseLU>
+
+#include <cmath>
+#include <iostream>
+#include <limits>
+#include <memory>
+#include <sstream>
+#include <vector>
+
+#include "Optimization.h"
+
+namespace {
+template <typename Derived>
+void selectSubMat(const Eigen::MatrixBase<Derived>& in,
+                  const std::vector<unsigned>& rows,
+                  const std::vector<unsigned>& cols,
+                  Eigen::MatrixXd* out) {
+  const int num_rows = rows.size();
+  const int num_cols = cols.size();
+  out->resize(num_rows, num_cols);
+
+  for (int i = 0; i < num_rows; i++) {
+    const auto row_in = in.row(rows[i]);
+    auto row_out = out->row(i);
+    for (int j = 0; j < num_cols; j++) {
+      row_out(j) = row_in(cols[j]);
+    }
+  }
+}
+
+// TODO sammy this could also use a more efficient implementation.
+template <typename Derived>
+void selectSubVec(const Eigen::MatrixBase<Derived>& in,
+                  const std::vector<unsigned>& rows,
+                  Eigen::VectorXd* out) {
+  const int num_rows = rows.size();
+  out->resize(num_rows);
+  for (int i = 0; i < num_rows; i++) {
+    (*out)(i) = in(rows[i]);
+  }
+}
+
+template <typename DerivedA, typename DerivedB, typename F>
+void transformVecElements(const Eigen::MatrixBase<DerivedA>& in,
+                          Eigen::MatrixBase<DerivedB>* out,
+                          F func) {
+  assert(in.cols() == 1);
+  assert(out->cols() == 1);
+  assert(in.rows() == out->rows());
+  for (int i = 0; i < in.rows(); i++) {
+    (*out)(i) = func(in(i), (*out)(i));
+  }
+}
+
+template <typename Derived>
+Eigen::SparseVector<double> makeSparseVector(
+    const Eigen::MatrixBase<Derived>& in) {
+  assert(in.cols() == 1);
+  Eigen::SparseVector<double> out(in.rows());
+  for (int i = 0; i < in.rows(); i++) {
+    if (in(i) != 0.0) {
+      out.coeffRef(i) = in(i);
+    }
+  }
+  return out;
+}
+
+template <typename Derived>
+Eigen::Index minCoeffIdx(const Eigen::MatrixBase<Derived>& in) {
+  Eigen::Index idx;
+  in.minCoeff(&idx);
+  return idx;
+}
+
+const double NEAR_ZERO = std::sqrt(std::numeric_limits<double>::epsilon());
```

Comment:
Sherm's note is correct. epsilon^2 (~1e-32) would also still be a representable result, in case you're not up on your IEEE 754 specs.

________________________________
> Date: Sun, 20 Mar 2016 13:28:58 -0700 
> From: notifications@github.com 
> To: drake@noreply.github.com 
> CC: edrumwri@hotmail.com 
> Subject: Re: [drake] Moby lcp integration (#1874) 
> 
> 
> In 
> drake/solvers/MobyLCP.cpp<https://github.com/RobotLocomotion/drake/pull/1874#discussion_r56771378>: 
> 
>> + for (int i = 0; i < in.rows(); i++) { 
>> + if (in(i) != 0.0) { 
>> + out.coeffRef(i) = in(i); 
>> + } 
>> + } 
>> + return out; 
>> +} 
>> + 
>> +template <typename Derived> 
>> +Eigen::Index minCoeffIdx(const Eigen::MatrixBase<Derived>& in) { 
>> + Eigen::Index idx; 
>> + in.minCoeff(&idx); 
>> + return idx; 
>> +} 
>> + 
>> +const double NEAR_ZERO = std::sqrt(std::numeric_limits<double>::epsilon()); 
> 
> 
> my read of 
> http://en.cppreference.com/w/cpp/types/numeric_limits/epsilon is that 
> this should not be able to evaluate to anything smaller than epsilon? 
> 
> � 
> You are receiving this because you were mentioned. 
> Reply to this email directly or view it on 
> GitHub<https://github.com/RobotLocomotion/drake/pull/1874/files/7ba1c61c5c67d54c7abe5b17f27d04edc08002f3#r56771378>